build/toml-patch/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+                        Value::Array(u) => {
+                            u.extend(v.as_array().unwrap().iter().cloned());
+                        }
+                    }
+                }
+                Item::Table(u) => {
+                    // Recurse!
+                    merge_toml_tables(u, v.as_table_mut().unwrap())?;
+                }
+                Item::ArrayOfTables(arr) => {
+                    // Compute an offset based on table position
+                    let mut visitor = TableRangeVisitor::default();
+                    visitor.visit_array_of_tables(arr);
+                    let range_in = visitor.range.unwrap();
+                    let last = range_in.end;
+
+                    let mut visitor = TableRangeVisitor::default();
+                    visitor.visit_item(v);
+                    let start =
+                        visitor.range.map(|r| r.start as isize).unwrap();
+                    let offset = last as isize - start as isize;
+
+                    // Apply that offset to the incoming tables
+                    let mut visitor = TableShiftVisitor { offset };
+                    visitor.visit_item_mut(v);
+
+                    // Merge by extending the table array
+                    arr.extend(v.as_array_of_tables().unwrap().iter().cloned());
+                }
+            }
+        } else {
+            let mut visitor = TableRangeVisitor::default();
+            visitor.visit_table(original);
+            let last = visitor.range.unwrap().end;
+
+            let mut visitor = TableRangeVisitor::default();
+            visitor.visit_item(v);
+            let start = visitor.range.map(|r| r.start as isize).unwrap_or(0);
+            let offset = last as isize - start as isize;
+
+            // Apply that offset to the incoming tables
+            let mut visitor = TableShiftVisitor { offset };
+            visitor.visit_item_mut(v);
+
+            // Merge by inserting the new element
+            original.insert(k.get(), v.clone());
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use indoc::indoc;
+
+    fn patch_and_compare(a: &str, b: &str, out: &str) {
+        let mut a: toml_edit::Document = a.parse().unwrap();
+        let b = b.parse().unwrap();
+        merge_toml_documents(&mut a, b).unwrap();
+        if a.to_string() != out {
+            eprintln!("patching failed.  Got result:");
+            eprintln!("{}", a.to_string());
+            eprintln!("----------------");
+            eprintln!("{}", out);
+        }
+        assert_eq!(a.to_string(), out);
+    }
+    #[test]
+    fn test_patching() {
+        patch_and_compare(
+            indoc! {r#"
+                name = "foo"
+                age = 37
+            "#},
+            indoc! {r#"
+                age = 38
+            "#},
+            indoc! {r#"
+                name = "foo"
+                age = 38
+            "#},
+        );
+        patch_and_compare(
```

Comment:
any reason that these are all in one big test instead of each being their own test? im guessing this runs fast enough that like, it doesn't *really* matter either way, I'm just curious :)